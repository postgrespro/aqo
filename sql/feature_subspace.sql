-- This test related to some issues on feature subspace calculation

CREATE EXTENSION IF NOT EXISTS aqo;
SELECT true AS success FROM aqo_reset();

SET aqo.mode = 'learn';
SET aqo.show_details = 'on';

CREATE TABLE a AS (SELECT gs AS x FROM generate_series(1,10) AS gs);
CREATE TABLE b AS (SELECT gs AS x FROM generate_series(1,100) AS gs);

--
-- Returns string-by-string explain of a query. Made for removing some strings
-- from the explain output.
--
CREATE OR REPLACE FUNCTION expln(query_string text) RETURNS SETOF text AS $$
BEGIN
    RETURN QUERY
        EXECUTE format('EXPLAIN (ANALYZE, COSTS OFF, TIMING OFF, SUMMARY OFF) %s', query_string);
    RETURN;
END;
$$ LANGUAGE PLPGSQL;

--
-- A LEFT JOIN B isn't equal B LEFT JOIN A.
--
SELECT str AS result
FROM expln('
SELECT * FROM a LEFT JOIN b USING (x);') AS str
WHERE str NOT LIKE '%Memory%';

-- TODO: Using method of other classes neighbours we get a bad estimation.
SELECT str AS result
FROM expln('
SELECT * FROM b LEFT JOIN a USING (x);') AS str
WHERE str NOT LIKE '%Memory%';

-- Look into the reason: two JOINs from different classes have the same FSS.
SELECT to_char(d1.targets[1], 'FM999.00') AS target FROM aqo_data d1
JOIN aqo_data d2 ON (d1.fs <> d2.fs AND d1.fss = d2.fss)
WHERE 'a'::regclass = ANY (d1.oids) AND 'b'::regclass = ANY (d1.oids) order by target;

DROP TABLE a,b CASCADE;

DROP EXTENSION aqo;
