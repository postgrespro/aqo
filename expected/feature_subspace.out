-- This test related to some issues on feature subspace calculation
CREATE EXTENSION IF NOT EXISTS aqo;
SELECT true AS success FROM aqo_reset();
 success 
---------
 t
(1 row)

SET aqo.mode = 'learn';
SET aqo.show_details = 'on';
CREATE TABLE a AS (SELECT gs AS x FROM generate_series(1,10) AS gs);
CREATE TABLE b AS (SELECT gs AS x FROM generate_series(1,100) AS gs);
--
-- Returns string-by-string explain of a query. Made for removing some strings
-- from the explain output.
--
CREATE OR REPLACE FUNCTION expln(query_string text) RETURNS SETOF text AS $$
BEGIN
    RETURN QUERY
        EXECUTE format('EXPLAIN (ANALYZE, COSTS OFF, TIMING OFF, SUMMARY OFF) %s', query_string);
    RETURN;
END;
$$ LANGUAGE PLPGSQL;
--
-- A LEFT JOIN B isn't equal B LEFT JOIN A.
--
SELECT str AS result
FROM expln('
SELECT * FROM a LEFT JOIN b USING (x);') AS str
WHERE str NOT LIKE '%Memory%';
                       result                        
-----------------------------------------------------
 Merge Left Join (actual rows=10 loops=1)
   AQO not used
   Merge Cond: (a.x = b.x)
   ->  Sort (actual rows=10 loops=1)
         Sort Key: a.x
         ->  Seq Scan on a (actual rows=10 loops=1)
               AQO not used
   ->  Sort (actual rows=11 loops=1)
         Sort Key: b.x
         ->  Seq Scan on b (actual rows=100 loops=1)
               AQO not used
 Using aqo: true
 AQO mode: LEARN
 JOINS: 0
(14 rows)

-- TODO: Using method of other classes neighbours we get a bad estimation.
SELECT str AS result
FROM expln('
SELECT * FROM b LEFT JOIN a USING (x);') AS str
WHERE str NOT LIKE '%Memory%';
                       result                        
-----------------------------------------------------
 Merge Left Join (actual rows=100 loops=1)
   AQO not used
   Merge Cond: (b.x = a.x)
   ->  Sort (actual rows=100 loops=1)
         Sort Key: b.x
         ->  Seq Scan on b (actual rows=100 loops=1)
               AQO not used
   ->  Sort (actual rows=10 loops=1)
         Sort Key: a.x
         ->  Seq Scan on a (actual rows=10 loops=1)
               AQO not used
 Using aqo: true
 AQO mode: LEARN
 JOINS: 0
(14 rows)

-- Look into the reason: two JOINs from different classes have the same FSS.
SELECT to_char(d1.targets[1], 'FM999.00') AS target FROM aqo_data d1
JOIN aqo_data d2 ON (d1.fs <> d2.fs AND d1.fss = d2.fss)
WHERE 'a'::regclass = ANY (d1.oids) AND 'b'::regclass = ANY (d1.oids) order by target;
 target 
--------
 2.30
 4.61
(2 rows)

DROP TABLE a,b CASCADE;
DROP EXTENSION aqo;
